### ISSUE TITLE
Attestation Agent should run before the pull_image function in image_rpc.rs for provisioning secrets using get_resource API for establishing a secure channel (Issue #53)

### ISSUE THREAD

**Issue by salmanyam** [confidential-containers/community#79]
To mitigate the attack on the kata control plane, issues [#1834](https://github.com/kata-containers/kata-containers/issues/1834) and [#53](https://github.com/confidential-containers/community/issues/53) discuss the idea of splitting the kata control plane into two parts for confidential containers by restricting the `kata-agent` API endpoints from the untrustworthy kata control plane and introducing another secure channel for accessing sensitive the `kata-agent` APIs.  However, to establish a secure channel (e.g., TLS) between the tenant and a secure server in the `kata-agent`, we need to provision secrets between them.

We can leverage the trusted KBS (relying party) to establish an attested TLS secure channel with the attestation agent within the confidential VM. This enables us to utilize the `get_resource` API of the `attestation-agent` to securely provision secrets onto the confidential VM, facilitating the setup of a secure channel with the tenant management stack.

In the current implementation of the `kata-agent`, the `attestation-agent` is instantiated once the `pull_image()` function in `image_rpc.rs` is called. This delay in the attestation agent startup restricts the usage of the `get_resource` API before the `pull_image()` function call. Thus, to support the split host/tenant APIs and potentially other use cases, it is necessary to ensure the startup of the attestation agent before the `pull_image()` function call.

In addition, the current code for starting the `attestation-agent` is embedded inside the `image_rpc.rs` file, perhaps due to the `pull_image()` function only utilizing it. However, a better idea is to refactor the initialization code of the `attestation-agent` from `image_rpc.rs` to move out the code to a separate file for a separate service.

We look forward to your response to this issue and the recommended changes.

**Comment by jepio** [comment]
Multiple thoughts:

What resources specifically would you like to fetch and when in the guest VM lifetime? Let's not be too abstract here :)

Do you want to make this mandatory? If so, how will the identifiers of needed resources be provided to the guest VM?

"Move out the code to a seperate file for a separate service" - why a new service? Or is this a matter of kata-agent taking a direct dependency on attestation-agent, instead of indirectly through image-rs?

Starting `attestation-agent` earlier seems sane. I would even suggest we start it as a systemd.service when systemd is available in the guest VM (cc: @mkulke)

Also: there is precedent for something related already, the KBS can provide the `policy.json` file right @fitzthum.

**Comment by mkulke** [comment]
So, I don't know the exact reasoning for running Attestation Agent (AA) as a discrete grpc service in the VM, alongside kata-agent. Maybe the purpose is to enable Pods to retrieve sources at runtime, maybe it's required to facilitate some back and forth between containerd fork/image-rs/ocicrypt-rs/kata-agent.

But indendept from the grpc service, you could use attestation service programmatically at any given time like this:

```rust
use attestation_agent::AttestationAPIs;
use attestation_agent::AttestationAgent;

let mut attestation_agent = AttestationAgent::new();
let resource_bytes = attestation_agent
    .download_confidential_resource(CC_KBC_NAME, &cli.resource_path, &cli.kbs_uri)
    .await?;
```

see [here](https://github.com/confidential-containers/kbs/blob/main/tools/client/src/main.rs).

Is this something that would work?

**Comment by salmanyam** [comment]
**_What resources specifically would you like to fetch and when in the guest VM lifetime?_**
-- We want to fetch TLS keys from KBS for setting up a gRPC server that will run side by side with the ttRPC server.
-- We want to start the gRPC server right after where the ttRPC server is started. The ttRPC server is started at line around [350](https://github.com/confidential-containers/kata-containers-CCv0/blob/CCv0/src/agent/src/main.rs#L350) in `main.rs` as follows.

```
// vsock:///dev/vsock, port
let mut server = rpc::start(sandbox.clone(), config.server_addr.as_str())?;
server.start().await?;
```

The following code is something we need for the split tenancy right after where the ttRPC server starts.
```
// We want to fetch the TLS keys here from KBS through the attestation agent, e.g.,
retrieve_secrets().await?;

// we want to start the gRPC server here e.g.,
let gserver = rpc::rpctls::grpcstart(sandbox.clone(), config.server_addr.as_str())?;
gserver.await?;
```

**_"Move out the code to a seperate file for a separate service" - why a new service?_**
-- I was talking more on on-demand control of initializing the attestation agent by moving all the attestation agent initialization code to a separate file. However, currently the initialization can only be done in `image_rps.rs` at [line 168](https://github.com/confidential-containers/kata-containers-CCv0/blob/CCv0/src/agent/src/image_rpc.rs#L168) using the `init_attestation_agent` function call at [line 90](https://github.com/confidential-containers/kata-containers-CCv0/blob/CCv0/src/agent/src/image_rpc.rs#L90).

We are proposing something as follows. For example, the split tenancy requires the initialization of the attestation agent before the pull_image() function call. So, we can do something like the following after where the ttRPC server is started ([line 350](https://github.com/confidential-containers/kata-containers-CCv0/blob/CCv0/src/agent/src/main.rs#L350)).

```
// obtaining a lock for the attestation agent and starting the agent if not started
let aa_init = aa_instance.lock().await;
aa_init.start_attestation_agent().await?;
//dropping the lock here
```
This will allow the pull_image() function to start the attestation agent if its not already started by obtaining the lock and calling `the start_attestation_agent()` function.


**_Is this something that would work? Regarding the example code_**
-- To the best of my understanding, I believe the example code assumes that the attestation initialization is already done. It will work if the AA is already running. If the AA is not running, `AttestationAgent::new()` will probably fail.

**Comment by mkulke** [comment]
> -- To the best of my understanding, I believe the example code assumes that the attestation initialization is already done. It will work if the AA is already running. If the AA is not running, the following will probably fail.

I don't think so, this is the sample client from the kbs repo. afaik an attestation-agent grpc app just wraps the attestation-agent library, it should do the same thing.

**Comment by Xynnn007** [comment]
> Starting `attestation-agent` earlier seems sane. I would even suggest we start it as a systemd.service when systemd is available in the guest VM (cc: @mkulke)

I prefer this way. To my understanding, Attestation-Agent is a "stub service", which can provide some services based on remote attestation for all containers in the guest VM, like

1. Get confidential resource --- By call the gRPC/ttrpc `get_resource`
2. Decrypt Image --- By call the gRPC/ttrpc `un_wrap_key`
3. serve as a "control-service" to receive commands from the relying party -- I think this is what split-api requires
4. get remote attestation report/token by containers -- I think this is what https://github.com/confidential-containers/kbs/issues/20 refer to
